### PR TITLE
feat: Adding new searcher that uses contributed surrogate models. 

### PR DESCRIPTION
**Pull request by jgolebiowski**
Adding new searcher that uses contributed surrogate models.
This comes with a new test and an example showing how it can be done

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: syne_tune/optimizer/schedulers/searchers/bayesopt/sklearn/contributed_surrogate_searcher.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+)
+from syne_tune.optimizer.schedulers.searchers.model_based_searcher import (
+    BayesianOptimizationSearcher,
+)
+from syne_tune.optimizer.schedulers.searchers.utils import make_hyperparameter_ranges
+
+logger = logging.getLogger(__name__)
+
+
+class ContributedSurrogateSearcher(BayesianOptimizationSearcher):
```
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/bayesopt/sklearn/contributed_surrogate_searcher.py:43]: This is very nice!

Once we have particular instances of this searcher, for example the ICML thing, we should then also invest in a factory, so user just has to specify `searcher` and `search_options` in the scheduler. But for now, this is exactly what we need.

File: examples/launch_contributed_surrogate_bo.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        "height": randint(1, 20),
+        "epochs": 10,
+    }
+    # Scheduler (i.e., HPO algorithm)
+    myestimator = BayesianRidgeEstimator()
+    searcher = ContributedSurrogateSearcher(
+        config_space=config_space,
+        metric="mean_loss",
+        estimator=myestimator,
+        scoring_class_and_args=IndependentThompsonSampling,
```
Comment (mseeger) [inline examples/launch_contributed_surrogate_bo.py:102]: Careful! This is precisely not what you want here! For ITS, this does not even use the acquisition function.

We should not allow the choice of `initial_scoring` for this searcher, because ITS does not make sense, and the only other thing is to use the AF for scoring.

File: syne_tune/optimizer/schedulers/searchers/bayesopt/sklearn/contributed_surrogate_searcher.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    Most of the implementation is generic in
+    :class:`~syne_tune.optimizer.schedulers.searchers.model_based_searcher.BayesianOptimizationSearcher`.
+    """
+
+    def __init__(
+        self,
+        config_space: Dict[str, Any],
+        metric: str,
+        estimator: SklearnEstimator,
+        scoring_class_and_args: ScoringClassAndArgs = None,
```
Comment (mseeger) [inline syne_tune/optimizer/schedulers/searchers/bayesopt/sklearn/contributed_surrogate_searcher.py:59]: This should be `acquisition_class`.

To explain this. In GP-BO, you can use the AF for (1) scoring initial cands, and (2) fine-tune by local search. But for (1), you can also use something else. For example, ITS is *independent* of the AF.

Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/bayesopt/sklearn/contributed_surrogate_searcher.py:59]: Sure we need `ScoringClassAndArgs`? I'd just use `AcquisitionClassAndArgs`.

File: examples/launch_contributed_surrogate_bo.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+from syne_tune.optimizer.schedulers.searchers.bayesopt.sklearn.contributed_surrogate_searcher import (
+    ContributedSurrogateSearcher,
+)
+from syne_tune.optimizer.schedulers.searchers.bayesopt.sklearn.estimator import (
+    SklearnEstimator,
+)
+from syne_tune.optimizer.schedulers.searchers.bayesopt.sklearn.predictor import (
+    SklearnPredictor,
+)
+from syne_tune.optimizer.schedulers.searchers.bayesopt.tuning_algorithms.bo_algorithm_components import (
```
Comment (mseeger) [inline examples/launch_contributed_surrogate_bo.py:33]: Don't use this. Sorry this is a bit confusing, see below

File: tst/schedulers/test_schedulers_api.py [hunk 5]
```diff
@@ -376,6 +382,27 @@ def make_transfer_learning_evaluations(num_evals: int = 10):
         base_scheduler_factory=FIFOScheduler,
         searcher="random",
     ),
+    FIFOScheduler(
+        config_space,
+        searcher=ContributedSurrogateSearcher(
+            config_space=config_space,
+            metric="mean_loss",
+            estimator=TestEstimator(),
+            scoring_class_and_args=IndependentThompsonSampling,
```
Comment (mseeger) [inline tst/schedulers/test_schedulers_api.py:391]: No, don't use this

File: tst/util_test.py [hunk 6]
```diff
@@ -35,6 +43,19 @@
 )
 
 
+class TestPredictor(SklearnPredictor):
```
Comment (mseeger) [inline tst/util_test.py:46]: Do you have an example somewhere, which also outputs predictive variances?

**Review by mseeger — CHANGES_REQUESTED** [review]

File: examples/launch_contributed_surrogate_bo.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+            * "mean": Predictive means in shape of ``(n,)``
+            * "std": Predictive stddevs, shape ``(n,)``
+        """
+        print(f"Predicting with BayesianRidgePredictor using X.shape={X.shape}")
+        return self.ridge.predict(X, return_std=True)
+
+
+class BayesianRidgeEstimator(SklearnEstimator):
+    """
+    Base class for the sklearn Estimators
```
Comment (wesk) [inline examples/launch_contributed_surrogate_bo.py:65]: `SklearnEstimator` is the base class, not `BayesianRidgeEstimator`, right?

File: syne_tune/optimizer/schedulers/searchers/bayesopt/sklearn/contributed_surrogate_searcher.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (wesk) [reply syne_tune/optimizer/schedulers/searchers/bayesopt/sklearn/contributed_surrogate_searcher.py:1]: I also like the specificity of `sklearn` and not `contributed`, which I find confusing

File: examples/launch_contributed_surrogate_bo.py (hunk 7 above)
Comment (jgolebiowski) [reply examples/launch_contributed_surrogate_bo.py:65]: yes, `BayesianRidgeEstimator` is only defined here for the purpose of the example

File: syne_tune/optimizer/schedulers/searchers/bayesopt/sklearn/contributed_surrogate_searcher.py (hunk 3 above)
Comment (jgolebiowski) [reply syne_tune/optimizer/schedulers/searchers/bayesopt/sklearn/contributed_surrogate_searcher.py:59]: I have added a the `ScoringClassAndArgs` in `syne_tune/optimizer/schedulers/searchers/bayesopt/tuning_algorithms/base_classes.py`. Overall I think the skleanr searcher will rely on acquisition functions w.o gradients. The local optimizatin is already being skipped in https://github.com/awslabs/syne-tune/pull/684/files#diff-63315ef366b2e2b3cde9ea92943d6c9e8db65fcd630e8f4fce379f754fdc406eR99 so I dont think the gradients would be useful.

Note that `AcquisitionFunction` inherits from`ScoringFunction` so any acquitision function can be used here. See the launch example that uss EI.

File: tst/schedulers/test_schedulers_api.py (hunk 5 above)
Comment (jgolebiowski) [reply tst/schedulers/test_schedulers_api.py:391]: changed to `LCBAcquisitionFunction`

File: tst/util_test.py (hunk 6 above)
Comment (jgolebiowski) [reply tst/util_test.py:46]: The example with a proper model is in launch examples. I wanted to keep this as simple as possible but updated the stds to be non-zero to make sure it does not crash anything

**Review by mseeger — APPROVED** [review]
Just a small thing: I'd write SKLearn instead of Sklearn. Everywhere.

**Review by mseeger — APPROVED** [review]
